enhancements/authentication/adding-uid-extra-claim-mappings-to-oidc.md

```diff
@@ -0,0 +1,417 @@
+- `uid.expression` - a CEL expression that results in the string value that should be used as the cluster user identity's UID.
+
+It will have the same validations enforced, at admission time where possible, as the Kubernetes API server enforces:
+- https://github.com/kubernetes/kubernetes/blob/b15dfce6cbd0d5bbbcd6172cf7e2082f4d31055e/staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go#L327-L340
+- CEL expression compilation will be enforced on HyperShift CRDs at runtime. On OpenShift, an admission webhook or plugin will be created to reject CEL expressions that won't compile at admission time.
+
+Some additional validations may be enforced, unique to OpenShift such as:
+- Requiring either `uid.claim` or `uid.expression` to be set when the `uid` field is specified.
+- Enforcing minimum and maximum length restrictions
+
+In practice, configuring these options would look something like:
+```yaml
+type: OIDC
+oidcProviders:
+- name: foo
+  claimMappings:
+    uid:
+     claim: email
+```
+or
+
+```yaml
+type: OIDC
+oidcProviders:
+- name: foo
+  claimMappings:
+    uid:
+     expression: claims.email
+```
+
+**NOTE:** These examples intentionally only include the `spec` fields that would need to be set to represent how it would look across all 3 APIs that would be modified.
+
+#### Adding the `extra` field
+The `extra` field will be added as an optional field.
+
+It will have 2 subfields:
+- `extra.key` - a string value to specify the key to be used for the extra attribute on the cluster user identity.
+- `extra.valueExpression` - a CEL expression that results in the string value that should be used as the value for the extra attribute.
+
+It will have the same validations enforced, at admission time where possible, as the Kubernetes API server enforces:
+- https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go#L345-L384
+- CEL expression compilation will be enforced on HyperShift CRDs at runtime. On OpenShift, an admission webhook or plugin will be created to reject CEL expressions that won't compile at admission time.
+
+Some additional validations may be enforced, unique to OpenShift such as:
+- Enforcing minimum and maximum length restrictions
+- Adding additional reserved domains and subdomains for OpenShift
+
+In practice, configuring these options would look something like:
+```yaml
+type: OIDC
+oidcProviders:
+- name: foo
+  claimMappings:
+    extra:
+    - key: example.org/foo
+      valueExpression: claims.foo
+```
+
+**NOTE:** This example intentionally only includes the `spec` fields that would need to be set to represent how it would look across all 3 APIs that would be modified.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+Failure mode communications:
+- Admission time validations will fail upon attempting to create a `HostedCluster` with invalid values specified for the new configuration options.
```

Comment:
HCP doesn't allow admission webhooks and cannot rely on kube-apiserver patches.  How will this be done?

Reply:
Standard admission time validations for a CRD.

Reply:
> Standard admission time validations for a CRD.

There are no standard admission checks for a CR instance.

Reply:
By standard admission checks, I mean things like writing CEL validations, min/max length, etc.

Is that insufficient? Which terminology should I use to represent that?